### DISCUSSION TITLE
How do I list all the undefined symbols in my playground wasm?

### DISCUSSION THREAD

**Discussion by ahelwer** [tree-sitter/tree-sitter#1325] [Q&A]
This might convert into an issue but we'll start with a discussion first. Related:
- https://github.com/tree-sitter/tree-sitter/issues/949
- https://github.com/tree-sitter/tree-sitter/issues/1300

After making some changes to my grammar's external scanner, the playground will crash when exercising certain external scanner code with the following error:
> Uncaught (in promise) TypeError: n is undefined
    t https://tlaplus-community.github.io/tree-sitter-tlaplus/js/tree-sitter.js:1
    parse https://tlaplus-community.github.io/tree-sitter-tlaplus/js/tree-sitter.js:1
    handleCodeChange https://tlaplus-community.github.io/tree-sitter-tlaplus/js/playground.js:115
    it https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.45.0/codemirror.min.js:1
    jn https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.45.0/codemirror.min.js:1
    Gn https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.45.0/codemirror.min.js:1
    Gn https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.45.0/codemirror.min.js:1
    Gn https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.45.0/codemirror.min.js:1
    Gn https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.45.0/codemirror.min.js:1
    Xn https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.45.0/codemirror.min.js:1
    poll https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.45.0/codemirror.min.js:1
    init https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.45.0/codemirror.min.js:1
    tt https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.45.0/codemirror.min.js:1
    init https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.45.0/codemirror.min.js:1
    se https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.45.0/codemirror.min.js:1
    Cl https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.45.0/codemirror.min.js:1
    Cl https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.45.0/codemirror.min.js:1
    fromTextArea https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.45.0/codemirror.min.js:1
    <anonymous> https://tlaplus-community.github.io/tree-sitter-tlaplus/js/playground.js:40
    async* https://tlaplus-community.github.io/tree-sitter-tlaplus/js/playground.js:461

Since this is the second time I've run into issues with the playground, it's more of a request for debugging strategies than a request for help here. How can I interpret this error? What sorts of things could be causing it? Is there a way to step through the code locally and find the exception, then map it back to a location in the source code? I can't reproduce this issue anywhere except in the playground - all tests pass, I can parse a huge number of example source files, and when editing a file with neovim with my grammar being used for highlighting, I see no error pop up.

Edit: I narrowed down the cause (fairly excruciatingly - it takes over two minutes to rebuild my wasm!) to a call to `std::stoi` and some uses of `std::string`. Was inspired to do this after adding `printf` statements to try to debug this but then it just errored immediately upon hitting the `printf`s so I figured this is because of symbols that aren't being included. Is there any way to quickly find out which symbols in my wasm file are undefined?

**Comment by ahlinc** [comment]
@ahelwer Did you try to use the `llvm-objdump` tool like was described in the #949 issue?

**Comment by danielo515** [comment]
I was having a  wry similar problem. In my case it was all because stdio and printf. Once I removed those then the playground started to work. Is there any reason why does not work? Is there any defined macro value that I can use for conditional compilation when compiling to wasi?

**Reply by ahlinc** [reply]
As a workaround you can wrap your `printf`s in conditional statements like described in PRs mentioned there: https://github.com/tree-sitter/tree-sitter/discussions/1218#discussioncomment-1018632

**Reply by danielo515** [reply]
Unless I misunderstood the referenced issue, it will not work. You need conditional compilation to not include the stdio or printf methods at all. Not calling them isn't enough, they must be completely stripped from the code

**Reply by ahlinc** [reply]
Correct, you can define own wrappers around `printf` and other functions like:
```cpp
  void print_column(TSLexer* lexer) {
    #ifdef TREE_SITTER_INTERNAL_BUILD // Conditional compilation wrapper, the body will be empty unless compiled by `tree-sitter parse` under the hood machinery
      if (std::getenv("TREE_SITTER_DEBUG")) { // Second wrapper to control the scanner output by the `-d` option of the `tree-sitter parse` command
        std::cout << "col " << lexer->get_column(lexer) << std::endl;
      }
    #endif
  }
```
Above I've added comments to an example from: https://github.com/tree-sitter/tree-sitter/pull/1187

**Reply by danielo515** [reply]
That is perfect and exactly what I was looking for.
I understand that the provided flag is only defined when the `parse` command is used and not when you generate or compile the wasi, right? That's awesome

**Reply by ahlinc** [reply]
Exactly, it's defined only by the parse command when it compiles a grammar by itself.
